### ISSUE TITLE
Precompiled bladebit-1.2.2 crashes with Illegal instruction on Haswell CPU .vs. AVX512

### ISSUE THREAD

**Issue by bladeuserpi** [Chia-Network/bladebit#140]
```
1) Problem description
Precompiled bladebit-1.2.2 crashes with Illegal instruction on Haswell CPU

2) Output when it crashes
Creating 1 plots:
 Output path           : /plot12/
 Thread count          : 64
 Warm start enabled    : false
 Farmer public key     : xxx
 Pool contract address : xxx

System Memory: 500/503 GiB.
Memory required: 416 GiB.
Allocating buffers.
Generating plot 1 / 1: xxx

Running Phase 1
Illegal instruction (core dumped)

3) Analysis

3a)
dmesg shows: 
traps: bladebit[28531] trap invalid opcode ip:405580 sp:7ffddce72300 error:0 in bladebit[400000+fe000]

3b) Look Up ip:405580
objdump -d bladebit |grep -C3 405580:
  405572:	41 b9 00 e6 6f 00    	mov    $0x6fe600,%r9d
  405578:	45 31 c0             	xor    %r8d,%r8d
  40557b:	bf 7f 00 00 00       	mov    $0x7f,%edi
  405580:	62 52 7d 48 7b c0    	vpbroadcastw %r8d,%zmm8 ---> HERE it crashes
  405586:	4c 89 ce             	mov    %r9,%rsi
  405589:	31 c9                	xor    %ecx,%ecx
  40558b:	89 c8                	mov    %ecx,%eax

3c) So that is related to AVX
"vpbroadcastw" with _register_ operand is AVX512 instruction
see https://www.felixcloutier.com/x86/vpbroadcastb:vpbroadcastw:vpbroadcastd:vpbroadcastq
and https://www.felixcloutier.com/x86/vpbroadcast

3d) The CPU (Haswell) can only do AVX2
Model name:          Intel(R) Xeon(R) CPU E5-2698 v3 @ 2.30GHz

3e) Test if it runs on AVX512
Yes, on Skylake it runs fine:
Model name:          Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz

4) Conclusion
The precompiled binary probably was compiled on AVX512-capable CPU and
the "CMakeLists.txt" has compiler option "-march=native".
That causes "invalid opcode" on non-AVX512 CPUs.

@haroldb: Let me know if you confirm this analysis?
Also note Ivy Bridge (e.g. E5-2697v2) can only do AVX (not even AVX2),
so that should be taken into account if you decide to also support older CPUs
with the precompiled binary.

**Comment by harold-b** [comment]
Thanks for the details bug report. You are absolutely right in pointing out `-march=native`. It was not supposed to be there, I accidentally left it on when testing during the migration to CMake.

I'll prioritize a fix for this.

**Comment by bladeuserpi** [comment]
```
I also compared performance of these 2 executable on Xeon Gold machine
which again shows unexpected results !

- bladebit-1.2.2 precompiled 
- my self-compiled version (which I probably compiled on AVX machine), I believe that was bladebit-1.1.x 

1a) Running self-compiled AVX-Version on Xeon Gold
124044.12443.out:Finished plotting in 350.49 seconds (5.84 minutes).
124044.12443.out:Finished plotting in 344.27 seconds (5.74 minutes).
124044.12443.out:Finished plotting in 343.97 seconds (5.73 minutes).
124044.12443.out:Finished plotting in 345.64 seconds (5.76 minutes).
124044.12443.out:Finished plotting in 360.38 seconds (6.01 minutes).


1b) Running precompiled-1.2.2 AVX512-Version on Xeon Gold (same machine as 1a)
124977.14366.out:Finished plotting in 351.19 seconds (5.85 minutes).
124977.14366.out:Finished plotting in 348.18 seconds (5.80 minutes).
124977.14366.out:Finished plotting in 348.49 seconds (5.81 minutes).
124977.14366.out:Finished plotting in 352.21 seconds (5.87 minutes).
124977.14366.out:Finished plotting in 366.53 seconds (6.11 minutes).

2) Discussion of results:
The AVX512-version runs about 4-5 seconds SLOWER on Xeon Gold than the AVX-version (~1.1% performance loss)
That is unexpected as AVX512 should be faster 

3) Now that could be final conclusion, but  looking into details shows more aspects:
- Some parts of bladebit run faster with AVX512
- Other parts of bladebit run faster with AVX

3a) 2nd runs from 124044 aka AVX

   119	Finished F1 generation in 1.33 seconds.
   121	Finished F1 sort in 7.24 seconds.
   124	  Finished pairing L/R groups in 8.5250 seconds. Created 4294941671 pairs.
   127	  Finished computing Fx in 8.1830 seconds.
   143	  Finished sorting in 13.09 seconds.
   144	Finished forward propagating table 2 in 30.13 seconds.
   147	  Finished pairing L/R groups in 8.4670 seconds. Created 4294837444 pairs.
   150	  Finished computing Fx in 8.6370 seconds.
   152	  Finished sorting in 13.58 seconds.
   153	Finished forward propagating table 3 in 31.00 seconds.
   156	  Finished pairing L/R groups in 8.4410 seconds. Created 4294667958 pairs.
   159	  Finished computing Fx in 8.8220 seconds.
   161	  Finished sorting in 13.47 seconds.
   162	Finished forward propagating table 4 in 31.04 seconds.
   165	  Finished pairing L/R groups in 8.4570 seconds. Created 4294329627 pairs.
   168	  Finished computing Fx in 8.8040 seconds.
   170	  Finished sorting in 13.42 seconds.
   171	Finished forward propagating table 5 in 30.99 seconds.
   174	  Finished pairing L/R groups in 8.4600 seconds. Created 4293654823 pairs.
   177	  Finished computing Fx in 8.4870 seconds.
   179	  Finished sorting in 12.95 seconds.
   180	Finished forward propagating table 6 in 30.20 seconds.
   183	  Finished pairing L/R groups in 8.4330 seconds. Created 4292276761 pairs.
   186	  Finished computing Fx in 8.0270 seconds.
   187	Finished forward propagating table 7 in 16.77 seconds.
   188	Finished Phase 1 in 178.70 seconds.
   191	  Finished prunning table 6 in 0.34 seconds.
   193	  Finished prunning table 5 in 7.55 seconds.
   195	  Finished prunning table 4 in 7.20 seconds.
   197	  Finished prunning table 3 in 7.07 seconds.
   199	  Finished prunning table 2 in 7.02 seconds.
   200	Finished Phase 2 in 29.47 seconds.
   203	  Finished compressing tables 1 and 2 in 20.72 seconds
   206	  Finished compressing tables 2 and 3 in 21.52 seconds
   209	  Finished compressing tables 3 and 4 in 21.36 seconds
   212	  Finished compressing tables 4 and 5 in 21.69 seconds
   215	  Finished compressing tables 5 and 6 in 22.77 seconds
   218	  Finished compressing tables 6 and 7 in 27.23 seconds
   220	Finished Phase 3 in 135.30 seconds.
   223	  Finished writing P7 in 0.45 seconds.
   225	  Finished writing C1 table in 0.00 seconds.
   227	  Finished writing C2 table in 0.00 seconds.
   229	  Finished writing C3 table in 0.34 seconds.
   230	Finished Phase 4 in 0.80 seconds.
   231	Finished plotting in 344.27 seconds (5.74 minutes).

3b) 2nd run from 124977 aka AVX512

  119	Finished F1 generation in 1.30 seconds.
   121	Finished F1 sort in 7.19 seconds.
   124	  Finished pairing L/R groups in 10.3610 seconds. Created 4294868403 pairs.
   127	  Finished computing Fx in 7.7600 seconds.
   143	  Finished sorting in 12.96 seconds.
   144	Finished forward propagating table 2 in 31.39 seconds.
   147	  Finished pairing L/R groups in 10.1310 seconds. Created 4294770976 pairs.
   150	  Finished computing Fx in 8.0410 seconds.
   152	  Finished sorting in 13.49 seconds.
   153	Finished forward propagating table 3 in 31.97 seconds.
   156	  Finished   groups in 10.3180 seconds. Created 4294602734 pairs.
   159	  Finished computing Fx in 8.3330 seconds.
   161	  Finished sorting in 13.39 seconds.
   162	Finished forward propagating table 4 in 32.34 seconds.
   165	  Finished pairing L/R groups in 10.2840 seconds. Created 4294258051 pairs.
   168	  Finished computing Fx in 8.3100 seconds.
   170	  Finished sorting in 13.38 seconds.
   171	Finished forward propagating table 5 in 32.27 seconds.
   174	  Finished pairing L/R groups in 10.1400 seconds. Created 4293556648 pairs.
   177	  Finished computing Fx in 7.9690 seconds.
   179	  Finished sorting in 12.87 seconds.
   180	Finished forward propagating table 6 in 31.28 seconds.
   183	  Finished pairing L/R groups in 10.2930 seconds. Created 4292014409 pairs.
   186	  Finished computing Fx in 7.4670 seconds.
   187	Finished forward propagating table 7 in 18.06 seconds.
   188	Finished Phase 1 in 185.81 seconds.
   191	  Finished prunning table 6 in 0.34 seconds.
   193	  Finished prunning table 5 in 7.55 seconds.
   195	  Finished prunning table 4 in 7.24 seconds.
   197	  Finished prunning table 3 in 7.07 seconds.
   199	  Finished prunning table 2 in 7.02 seconds.
   200	Finished Phase 2 in 29.53 seconds.
   203	  Finished compressing tables 1 and 2 in 20.64 seconds
   206	  Finished compressing tables 2 and 3 in 21.39 seconds
   209	  Finished compressing tables 3 and 4 in 21.24 seconds
   212	  Finished compressing tables 4 and 5 in 21.59 seconds
   215	  Finished compressing tables 5 and 6 in 22.71 seconds
   218	  Finished compressing tables 6 and 7 in 24.51 seconds
   220	Finished Phase 3 in 132.09 seconds.
   223	  Finished writing P7 in 0.43 seconds.
   225	  Finished writing C1 table in 0.00 seconds.
   227	  Finished writing C2 table in 0.00 seconds.
   229	  Finished writing C3 table in 0.32 seconds.
   230	Finished Phase 4 in 0.75 seconds.
   231	Finished plotting in 348.18 seconds (5.80 minutes).


3c) Table with typical per-phase runtime in seconds
 
             AVX   AVX512
pairing L/R   8.5    10.3
computing Fx  8.5    7.8-8.3  # Computing Fx is ~0.5s faster with AVX512, 6 times =3s 
sorting       13.4    ~13.4
Phase 1      178      185
Phase2       29.5    29.5
Phase3      135      132    # Phase3 is 3s faster with AVX512
Phase4       0.8     0.75
Plot       344      348 

3) Future direction

Using the best of both could reduce the 344s runtime by another 6s to 338s (~2% potential gain)

Hint: bladebit contains blake3 which shows how a single executable can provide multiple
optimized routines(SSE/AVX/AVX512) which are selected at runtime.

**Comment by harold-b** [comment]
For people wanting the best performance, their best bet is to compile it themselves. The compiler will decide to vectorize certain parts of the code where it deems it beneficial. The other viable option would be to actually provide different variants of pre-compiled executables with the architectures explicitly specified. There's currently no portion of the BB code which does explicit SIMD, so one is not simply able to check at runtime for supported instructions to provide a faster version. This may be the case in some Phase3 portions later one where some explicit vectorization has already been prototyped.

**Comment by harold-b** [comment]
Thanks again for the report @bladeuserpi. This has been fixed in [v1.2.3](https://github.com/Chia-Network/bladebit/releases/tag/v1.2.3)
